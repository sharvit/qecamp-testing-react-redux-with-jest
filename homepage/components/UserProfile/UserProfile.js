/* UserProfile.js */
import React from 'react';

import UserAvatar from './components/UserAvatar';
import UserDetailsBox from './components/UserDetailsBox';
import UserPhotos from './components/UserPhotos';
import UserPosts from './components/UserPosts';

const UserProfile = ({
  user,
  showAvatar,
  showPosts,
  showPhotos,
  updateShowAvatar,
  updateShowPhotos,
  updateShowPosts,
}) => (
  <div className="user-profile">
    <UserDetailsBox user={user} />
    {showAvatar && <UserAvatar user={user} size="sm" />}
    {showPosts && <UserPhotos user={user} count={5} sort="DESC" />}
    {showPhotos && <UserPosts user={user} count={5} sort="DESC" />}
    <input
      id="show-avatar-toggler"
      type="checkbox"
      checked={showAvatar}
      onChange={e => updateShowAvatar(e.target.checked)}
    />
    <input
      id="show-photos-toggler"
      type="checkbox"
      checked={showPhotos}
      onChange={e => updateShowPhotos(e.target.checked)}
    />
    <input
      id="show-posts-toggler"
      type="checkbox"
      checked={showPosts}
      onChange={e => updateShowPosts(e.target.checked)}
    />
  </div>
);

export default UserProfile;
