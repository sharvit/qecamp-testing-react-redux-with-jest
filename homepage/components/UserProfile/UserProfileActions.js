/* UserProfileActions.js */
import {
  USER_PROFILE_UPDATE_AVATAR,
  USER_PROFILE_UPDATE_POSTS,
  USER_PROFILE_UPDATE_PHOTOS,
} from './UserProfileConstants';

export const updateShowAvatar = showAvatar => ({
  type: USER_PROFILE_UPDATE_AVATAR,
  payload: showAvatar,
});

export const updateShowPosts = showPosts => ({
  type: USER_PROFILE_UPDATE_POSTS,
  payload: showPosts,
});

export const updateShowPhotos = showPhotos => ({
  type: USER_PROFILE_UPDATE_PHOTOS,
  payload: showPhotos,
});
