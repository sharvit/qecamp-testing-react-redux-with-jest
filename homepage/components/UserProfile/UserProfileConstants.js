export const USER_PROFILE_UPDATE_AVATAR = 'USER_PROFILE_UPDATE_AVATAR';
export const USER_PROFILE_UPDATE_POSTS = 'USER_PROFILE_UPDATE_POSTS';
export const USER_PROFILE_UPDATE_PHOTOS = 'USER_PROFILE_UPDATE_PHOTOS';
