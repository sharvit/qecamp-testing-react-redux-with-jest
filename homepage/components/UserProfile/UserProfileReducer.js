import Immutable from 'seamless-immutable';

import {
  USER_PROFILE_UPDATE_AVATAR,
  USER_PROFILE_UPDATE_POSTS,
  USER_PROFILE_UPDATE_PHOTOS,
} from './UserProfileConstants';

const initialState = Immutable({
  showAvatar: false,
  showPosts: false,
  showPhotos: false,
});

export default (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case USER_PROFILE_UPDATE_AVATAR:
      return state.set('showAvatar', payload);
    case USER_PROFILE_UPDATE_POSTS:
      return state.set('showPosts', payload);
    case USER_PROFILE_UPDATE_PHOTOS:
      return state.set('showPhotos', payload);
    default:
      return state;
  }
};
