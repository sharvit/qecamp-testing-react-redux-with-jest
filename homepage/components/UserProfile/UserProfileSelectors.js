const selectState = state => state.userProfile;

export const selectShowAvatar = state => selectState(state).showAvatar;
export const selectShowPosts = state => selectState(state).showPosts;
export const selectShowPhotos = state => selectState(state).showPhotos;
