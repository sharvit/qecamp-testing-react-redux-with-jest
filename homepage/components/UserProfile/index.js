import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './UserProfileActions';
import reducer from './UserProfileReducer';
import {
  selectShowAvatar,
  selectShowPosts,
  selectShowPhotos,
} from './UserProfileSelectors';

import UserProfile from './UserProfile';

// map state to props
const mapStateToProps = state => ({
  showAvatar: selectShowAvatar(state),
  showPosts: selectShowPosts(state),
  showPhotos: selectShowPhotos(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { userProfile: reducer };

// export connected component
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserProfile);
