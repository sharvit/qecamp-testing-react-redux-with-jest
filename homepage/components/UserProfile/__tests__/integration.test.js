import React from 'react';
import IntegrationTestHelper from '../../../../IntegrationTestHelper';

import UserProfile, { reducers } from '../index';

describe('UserProfile - Integration Test', () => {
  it('should flow', () => {
    const integrationTestHelper = new IntegrationTestHelper(reducers);

    const component = integrationTestHelper.mount(
      <UserProfile user="some-user" />
    );

    // The user-avatar should not be shown
    expect(component.exists('UserAvatar')).toEqual(false);
    integrationTestHelper.takeStoreSnapshot('initial state');

    // trigger checkbox change
    component
      .find('input#show-avatar-toggler')
      .simulate('change', { target: { checked: true } });

    // The user-avatar should be shown now
    expect(component.exists('UserAvatar')).toEqual(true);
    integrationTestHelper.takeStoreAndLastActionSnapshot(
      'Update to show the user-avatar'
    );
  });
});
