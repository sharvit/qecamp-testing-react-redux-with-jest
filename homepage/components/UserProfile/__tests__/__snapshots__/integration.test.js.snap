// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile - Integration Test should flow: Update to show the user-avatar 1`] = `
Object {
  "action": Array [
    Array [
      Object {
        "payload": true,
        "type": "USER_PROFILE_UPDATE_AVATAR",
      },
    ],
  ],
  "state": Object {
    "userProfile": Object {
      "showAvatar": true,
      "showPhotos": false,
      "showPosts": false,
    },
  },
}
`;

exports[`UserProfile - Integration Test should flow: initial state 1`] = `
Object {
  "userProfile": Object {
    "showAvatar": false,
    "showPhotos": false,
    "showPosts": false,
  },
}
`;
